File: app/schemas/team.py
"""Team request/response schemas (Pydantic v2).

DESIGN:
- TeamOut carries denormalized read-time aggregates (member_count,
  supervisor_name, present_today, performance_pct) computed by the repository
  in one grouped query — the team-list card needs all of them, and computing
  them in SQL beats N round-trips.
- performance_pct = present_today / member_count * 100, rounded to 1 dp.
  member_count == 0 ⇒ 0.0 (no division by zero, an empty team is 0% not N/A).
- Soft delete: DELETE marks is_active=False (teams are referenced by users +
  audit history; a hard delete would orphan or cascade-null those).
"""
from pydantic import BaseModel, ConfigDict, Field


class TeamCreate(BaseModel):
    name: str = Field(min_length=2, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    supervisor_id: int | None = None


class TeamUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=2, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    supervisor_id: int | None = None


class AddMemberRequest(BaseModel):
    user_id: int


class TeamOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    supervisor_id: int | None
    supervisor_name: str | None
    member_count: int
    present_today: int
    performance_pct: float


class TeamMemberOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    email: str
    role: str
    profile_photo_url: str | None
    is_active: bool
    # Redis-derived live status (ACTIVE | IDLE | OFFLINE); read-time, never
    # persisted. Lets the Teams screen show who's live per team.
    live_status: str = "OFFLINE"


class TeamDetailOut(TeamOut):
    members: list[TeamMemberOut]

File: mobile/lib/features/teams/models/team.dart
import '../../auth/models/user.dart' show UserRole;

class TeamMember {
  const TeamMember({
    required this.id,
    required this.name,
    required this.email,
    required this.role,
    required this.isActive,
    this.profilePhotoUrl,
    this.liveStatus = 'OFFLINE',
  });

  final int id;
  final String name;
  final String email;
  final UserRole role;
  final bool isActive;
  final String? profilePhotoUrl;
  final String liveStatus; // ACTIVE | IDLE | OFFLINE (Redis-derived, read-time)

  String get initials {
    final parts = name.trim().split(RegExp(r'\s+')).where((p) => p.isNotEmpty);
    if (parts.isEmpty) return '?';
    return parts.take(2).map((p) => p[0].toUpperCase()).join();
  }

  factory TeamMember.fromJson(Map<String, dynamic> json) => TeamMember(
        id: json['id'] as int,
        name: json['name'] as String,
        email: json['email'] as String,
        role: UserRole.fromWire(json['role'] as String),
        isActive: (json['is_active'] as bool?) ?? true,
        profilePhotoUrl: json['profile_photo_url'] as String?,
        liveStatus: (json['live_status'] as String?) ?? 'OFFLINE',
      );
}

class Team {
  const Team({
    required this.id,
    required this.name,
    required this.memberCount,
    required this.presentToday,
    required this.performancePct,
    this.description,
    this.supervisorId,
    this.supervisorName,
    this.members = const [],
  });

  final int id;
  final String name;
  final int memberCount;
  final int presentToday;
  final double performancePct;
  final String? description;
  final int? supervisorId;
  final String? supervisorName;
  final List<TeamMember> members;

  factory Team.fromJson(Map<String, dynamic> json) => Team(
        id: json['id'] as int,
        name: json['name'] as String,
        memberCount: (json['member_count'] as int?) ?? 0,
        presentToday: (json['present_today'] as int?) ?? 0,
        performancePct: ((json['performance_pct'] as num?) ?? 0).toDouble(),
        description: json['description'] as String?,
        supervisorId: json['supervisor_id'] as int?,
        supervisorName: json['supervisor_name'] as String?,
        members: ((json['members'] as List<dynamic>?) ?? [])
            .map((e) => TeamMember.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
